client/v2/client.go

```diff
@@ -413,16 +414,45 @@ func (uc *udpclient) Write(bp BatchPoints) error {
 	for _, p := range bp.Points() {
 		pointstring := p.pt.RoundedString(d) + "\n"
 
-		// Write and reset the buffer if we reach the max size
 		if b.Len()+len(pointstring) >= uc.payloadSize {
 			if _, err := uc.conn.Write(b.Bytes()); err != nil {
 				return err
 			}
 			b.Reset()
 		}
 
-		if _, err := b.WriteString(pointstring); err != nil {
-			return err
+		if b.Len()+len(pointstring) <= uc.payloadSize {
+			if _, err := b.WriteString(pointstring); err != nil {
+				return err
+			}
+		} else {
+			if p.Time().IsZero() {
+				if len(pointstring) > MaxPointSize {
+					return fmt.Errorf("point size is greater than %dB", MaxPointSize)
+				}
+				if _, err := b.WriteString(pointstring); err != nil {
+					return err
+				}
+			} else {
+				points, err := p.pt.SplitN(uc.payloadSize - 2) // -2 because of `+ "\n"`
+				if err != nil {
+					return err
+				}
+				for _, sp := range points {
+					pointstring := sp.RoundedString(d) + "\n"
+
+					if b.Len()+len(pointstring) >= uc.payloadSize {
```

Comment:
This should also check to make sure `b.Len() > 0`, if `len(pointstring) == uc.payloadSize` and the buffer is empty, this will attempt a 0-length write to the connection.